src/entity/object/AreaEffectCloud.php

```diff
@@ -0,0 +1,416 @@
+		$this->duration = $duration;
+		$this->networkPropertiesDirty = true;
+	}
+
+	/**
+	 * Returns the amount that the duration of this cloud will add by when it
+	 * applies an effect to an entity.
+	 */
+	public function getDurationOnUse() : int{
+		return $this->durationOnUse;
+	}
+
+	/**
+	 * Sets the amount that the duration of this cloud will add by when it
+	 * applies an effect to an entity.
+	 */
+	public function setDurationOnUse(int $durationOnUse) : void{
+		$this->durationOnUse = $durationOnUse;
+		$this->networkPropertiesDirty = true;
+	}
+
+	/**
+	 * Returns the time that an entity will be immune from subsequent exposure.
+	 */
+	public function getReapplicationDelay() : int{
+		return $this->reapplicationDelay;
+	}
+
+	/**
+	 * Sets the time that an entity will be immune from subsequent exposure.
+	 */
+	public function setReapplicationDelay(int $delay) : void{
+		$this->reapplicationDelay = $delay;
+		$this->networkPropertiesDirty = true;
+	}
+
+	protected function entityBaseTick(int $tickDiff = 1) : bool{
+		$hasUpdate = parent::entityBaseTick($tickDiff);
+
+		$this->age += $tickDiff;
+		$radius = $this->radius + ($this->radiusPerTick * $tickDiff);
+		if($radius < 0.75){
+			$this->flagForDespawn();
+			return true;
+		}
+		$this->setRadius($radius);
+		if($this->age >= $this->waiting){
+			if($this->age > $this->duration){
+				$this->flagForDespawn();
+				return true;
+			}
+			foreach($this->victims as $entityId => $expiration){
+				if($this->age >= $expiration){
+					unset($this->victims[$entityId]);
+				}
+			}
+			foreach($this->getWorld()->getCollidingEntities($this->getBoundingBox(), $this) as $entity){
+				if(!$entity instanceof Living || isset($this->victims[$entity->getId()])){
+					continue;
+				}
+
+				$entityPosition = $entity->getPosition();
+				$xDiff = $entityPosition->getX() - $this->location->getX();
+				$zDiff = $entityPosition->getZ() - $this->location->getZ();
+				if(($xDiff ** 2 + $zDiff ** 2) > $this->radius ** 2){
+					continue;
+				}
+
+				foreach($this->getCloudEffects() as $effect){
+					if($effect->getType() instanceof InstantEffect){
+						$effect->getType()->applyEffect($entity, $effect, 0.5, $this);
+					}else{
+						$entity->getEffects()->add($effect->setDuration((int) round($effect->getDuration() / 4)));
+					}
+				}
+				if($this->reapplicationDelay !== 0){
+					$this->victims[$entity->getId()] = $this->age + $this->reapplicationDelay;
+				}
+				if($this->radiusOnUse !== 0.0){
+					$radius = $this->radius + $this->radiusOnUse;
+					if($radius <= 0){
+						$this->flagForDespawn();
+						return true;
+					}
+					$this->setRadius($radius);
+				}
+				if($this->durationOnUse !== 0){
+					$duration = $this->duration + $this->durationOnUse;
+					if($duration <= 0){
+						$this->flagForDespawn();
+						return true;
+					}
+					$this->setDuration($duration);
+				}
+			}
+			$this->setWaiting($this->age + self::WAIT_TIME);
```

Comment:
I think you're using this incorrectly. This will prevent the effect cloud from applying effects for 30 seconds at a time, which defeats the purpose of having `reapplication delay` of 2 seconds.

According to the wiki: The time before deploying the field. The Radius is ignored, meaning that any specified effects is not applied and specified particles appear only at the center of the field, until Age hits this number.

Reply:
I tested for this in vanilla, instant effects have a `reapplication Delay` of 0, but due to `wait Time` they are only applied every 10 ticks. So I noticed that `wait Time` is not only used during the first ticks of the entity's life. I also saw the java code and although the `wait Time` there is 5 ticks, `wait Time` also works as a delay for applying effects.

So...

`WaitTime`: Cloud Delay to apply effects, or update.
`ReapplicationDelay`: Delay per entity to reapply the effects, capped by `WaitTime`.